src/viam/sdk/common/exception.hpp

```diff
@@ -0,0 +1,52 @@
+/// @file common/exception.hpp
+///
+/// @brief Defines custom exceptions for the SDK.
+#pragma once
+#include <stdexcept>
+#include <string>
+
+#include <viam/sdk/resource/resource_api.hpp>
+
+namespace viam {
+namespace sdk {
+
+/// @defgroup Exception Classes related to SDK exceptions.
+
+/// @class ViamErrorCode
+/// @brief Defines a set of a error codes to be used in conjunction with
+/// ViamException.
+/// @ingroup Exception
+enum class ViamErrorCode {
+    Unknown = 0,            // Default code
+    Connection,             // Issue during connection establishment
+    DuplicateRegistration,  // API or API/Model pair has already been registered
+    DuplicateResource,      // Resource has already been added
+    GRPC,                   // gRPC error from remote machine
+    NotSupported,           // Behavior not supported by the SDK
+    ResourceNotFound        // Resource does not exist
+};
+
+std::error_code make_error_code(ViamErrorCode e);
```

Comment:
`std::error_code` or `std::error_condition`? I always struggle with the distinction.

Reply:
I _think_ `std::error_condition` is probably better here. The [distinction](https://en.cppreference.com/w/cpp/error/error_condition) seems to be that `std::error_code` is "platform-dependent" while `std::error_condition` is "platform-independent". I believe we want the latter.